drv/riscv-plic/build.rs

```diff
@@ -0,0 +1,163 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+use std::collections::HashMap;
+use std::env;
+use std::fs::File;
+use std::io::Write;
+use std::path::PathBuf;
+
+task_config::task_config! {
+    ints: &'static [u16],
+    tasks: &'static [&'static str],
+    notification: &'static [u32],
+    base: u32,
+    pbits: u8,
+}
+
+fn main() -> Result<(), Box<dyn std::error::Error>> {
+    build_util::expose_target_board();
+
+    idol::server::build_server_support(
+        "../../idl/riscv-plic.idol",
+        "server_stub.rs",
+        idol::server::ServerStyle::InOrder,
+    )?;
+
+    let out = PathBuf::from(env::var_os("OUT_DIR").unwrap());
+    let mut file = File::create(out.join("plic_config.rs")).unwrap();
+
+    writeln!(
+        file,
+        "use phash::{{PerfectHashMap, MutablePerfectHashMap}};"
+    )?;
+
+    writeln!(
+        file,
+        "const PLIC_PRIORITY_BITS: usize = 0x{:X};",
+        TASK_CONFIG.pbits
+    )
+    .unwrap();
+    writeln!(file, "const PLIC_REGISTER_BLOCK: *mut Plic<PLIC_PRIORITY_BITS> = 0x{:X} as *mut Plic<PLIC_PRIORITY_BITS>;", TASK_CONFIG.base).unwrap();
+    writeln!(file, "type Priority = plic::Priority<PLIC_PRIORITY_BITS>;")
+        .unwrap();
+
+    /*
+    let mut task_set: HashSet<String> = HashSet::new();
+    for task in TASK_CONFIG.tasks.iter() {
+        if task_set.insert(task.to_string()) == true {
+            writeln!(file, "task_slot!({}, {});", task.to_ascii_uppercase(), task);
+    }
+    */
+
+    use abi::{InterruptNum, InterruptOwner, TaskId};
+    let fmt_irq_task = |v: Option<&(InterruptNum, (TaskId, u32))>| {
+        match v {
+            Some((irq, owner)) => format!(
+                "(userlib::InterruptNum({}), (TaskId({}), 0b{:b})),",
+                irq.0, owner.0.0, owner.1
+            ),
+            None => "(userlib::InterruptNum::invalid(), userlib::InterruptOwner::invalid()),"
```

Comment:
This was lifted from `sys/kern/build.rs`, and I'm unsure why they had this as well. I'm guessing there is some situation where this can legally happen as they have other cases where the build will fail when making these maps. I can still have it fail instead if that's preferable.

Reply:
Yeah, I saw that. I was not sure why it was there. I think we can keep it as is for now (makes it easier to combine the code into a separate util crate).